src/main/java/ch/njol/skript/effects/EffToggle.java

```diff
@@ -40,63 +39,88 @@
 public class EffToggle extends Effect {
 	
 	static {
-		Skript.registerEffect(EffToggle.class, "(close|turn off|de[-]activate) %blocks%", "(toggle|switch) [[the] state of] %blocks%", "(open|turn on|activate) %blocks%");
+		Skript.registerEffect(EffToggle.class,
+			"(open|turn on|activate) %blocks%",
+			"(close|turn off|de[-]activate) %blocks%",
+			"(toggle|switch) [[the] state of] %booleans/blocks%");
+	}
+
+	private enum State {
+		ACTIVATE, DEACTIVATE, TOGGLE
 	}
 
 	@SuppressWarnings("null")
-	private Expression<Block> blocks;
-	private int toggle;
+	private Expression<?> togglables;
+	private State state;
 	
-	@SuppressWarnings({"unchecked", "null"})
 	@Override
-	public boolean init(final Expression<?>[] vars, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		blocks = (Expression<Block>) vars[0];
-		toggle = matchedPattern - 1;
+	public boolean init(Expression<?>[] vars, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		togglables = (Expression<?>) vars[0];
+		state = State.values()[matchedPattern];
+		if (togglables.getReturnType() == Boolean.class && !ChangerUtils.acceptsChange(togglables, ChangeMode.SET, Boolean.class)) {
+			Skript.error(togglables.toString(null, false) + " cannot be toggled");
+			return false;
+		}
 		return true;
 	}
 
 	@Override
-	protected void execute(final Event e) {
-		for (Block b : blocks.getArray(e)) {
-			BlockData data = b.getBlockData();
-			if (toggle == -1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(false);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(false);
-			} else if (toggle == 1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(true);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(true);
-			} else {
-				if (data instanceof Openable) // open = NOT was open
-					((Openable) data).setOpen(!((Openable) data).isOpen());
-				else if (data instanceof Powerable) // power = NOT power
-					((Powerable) data).setPowered(!((Powerable) data).isPowered());
+	protected void execute(Event event) {
+		ArrayList<Object> toggledValues = new ArrayList<>();
+		for (Object obj : togglables.getArray(event)) {
+			if (obj instanceof Block) {
+				Block block = (Block) obj;
+				BlockData data = block.getBlockData();
+				if (state == State.TOGGLE) {
+					if (data instanceof Openable) { // open = NOT was open
+						((Openable) data).setOpen(!((Openable) data).isOpen());
+					} else if (data instanceof Powerable) { // power = NOT power
+						((Powerable) data).setPowered(!((Powerable) data).isPowered());
+					}
+				} else {
+					boolean value = state == State.ACTIVATE; 
+					if (data instanceof Openable) {
+						((Openable) data).setOpen(value);
+					} else if (data instanceof Powerable) {
+						((Powerable) data).setPowered(value);
+					}
+				}
+
+				block.setBlockData(data);
+				toggledValues.add(block);
```

Comment:
I think this can be removed as I don't believe the block reference needs updated when changing the blockdata. Additionally, this means `toggledValues` would be a `List<Boolean>` so you can just check whether togglables accepts Boolean.class

Reply:
Not sure to get what you mean, but ``filteredValues`` is maybe not necessary, i even think it removes values (not boolean toggleable values), what do you think?